week-3/CrowdFunding/CrowdFunding.sol

```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.8.17;
+
+contract CrowdFunding {
+
+    // struct to hold contributor details
+    struct Contributor {
+        uint256 donated;
+        address donerId;
+    }
+
+    // struct to hold project details
+    struct Project {
+        uint256 projectId;
+        string projectName;
+        uint256 targetAmount;
+        uint256 collectedAmount;
+        address payable beneficiaryAddress;
+        Contributor[] contributors;
+        bool isOpen;
+    }
+
+    // state variables
+    address admin;
+    uint256 public projectIndex;
+    mapping(uint256 => Project) public projectList;
+
+    // events
+    event contributionAdded(
+        string indexed _projectName,
+        uint256 indexed _donation,
+        address indexed _donor
+    );
+    event fundingTargetAchieved(
+        string indexed _projectName,
+        uint256 indexed _amount,
+        address indexed _beneficiary
+    );
+    event fundsReturned(
+        string indexed _projectName,
+        uint256 indexed _refundAmount,
+        address indexed _refundAddress
+    );
+
+    // set admin address on contract deployment
+    constructor() {
+        admin = msg.sender;
+    }
+
+    // register new project for funding
+    function register(
+        string memory _name,
+        uint256 _target,
+        address payable _beneficiary
+    ) public {
+        Project storage newProject = projectList[projectIndex];
+
+        newProject.projectId = projectIndex;
+        newProject.projectName = _name;
+        newProject.targetAmount = _target;
+        newProject.collectedAmount = 0;
+        newProject.beneficiaryAddress = _beneficiary;
+        newProject.isOpen = true;
+
+        projectIndex++;
+    }
+
+    // return funds to contributors if target not achived
+    function returnFunds(
+        uint256 _projectId
+    ) public isAdmin isFundingOpen(_projectId) {
+        Project storage project = projectList[_projectId];
```

Comment:
Avoid making external calls like transfer in for loops as much as possible. Instead create a function where the contributors can withdraw their funds. https://fravoll.github.io/solidity-patterns/pull_over_push.html

Reply:
Check for Funding Achieved is missing.